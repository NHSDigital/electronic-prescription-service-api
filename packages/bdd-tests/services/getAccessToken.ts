/**
 * TypeScript version of `identity_service.py` from pytest-nhsd-apim
 * https://github.com/NHSDigital/pytest-nhsd-apim/blob/82316e098a8696e4f7b0de7747b580dc0ffa7229/
 * /src/pytest_nhsd_apim/identity_service.py#L298
 *
 * 1. Hit `authorize` endpoint w/ required query params --> we
 * are redirected to the simulated_auth page. The requests package
 * follows those redirects.
 *
 * 2. Parse the login page.  For keycloak this presents an
 * HTML form, which must be filled in with valid data.  The tester
 * can submits their login data with the `login_form` field.
 *
 * 3. POST the filled in form. This is equivalent to clicking the
 * "Login" button if we were a human.
 *
 * 4. The mock auth redirected us back to the
 * identity-service, which redirected us to whatever our app's
 * callback-url was set to.  We don't actually care about the
 * content our callback-url page, we just need the auth_code that
 * was provided in the redirect.
 *
 * 5. Finally, get an access token.
 *
 * 6. Profit... sweet sweet profit.
 */
import axios, {AxiosInstance, AxiosResponse} from "axios"
import {JSDOM} from "jsdom"
import {wrapper} from "axios-cookiejar-support"
import {CookieJar} from "tough-cookie"

export const VALID_APIGEE_ENVIRONMENTS = ["internal-dev", "internal-qa", "int", "ref"]

type EnvironmentSecrets = {
  clientId: string
  clientSecret: string
}

export class AuthClient {
  private readonly environment: string
  readonly clientId: string
  readonly clientSecret: string
  readonly callbackUrl: string
  private readonly state: string

  constructor() {
    const {clientId, clientSecret} = this.getApiCredentials()
    this.clientId = clientId
    this.clientSecret = clientSecret

    this.environment = this.getEnvironment()
    this.callbackUrl = "https://google.com" // using /callback causes the website to return a 404, which upsets axios
    this.state = this.getState()
  }

  getBaseUrl(): string {
    return `https://${this.environment}.api.service.nhs.uk/oauth2-mock`
  }

  getKeycloakUrl(): string {
    return `https://identity.ptl.api.platform.nhs.uk/auth/realms/Cis2-mock-internal-dev/protocol/openid-connect`
  }

  private getState(): string {
    // return uuid.v4()
    return "1234567890"
  }

  getEnvironment(): string {
    const env = process.env.APIGEE_ENVIRONMENT

    const isLocalEnv = process.env.NODE_ENV !== "production"
    const isValidEnv = VALID_APIGEE_ENVIRONMENTS.includes(env)
    if (!isLocalEnv || !isValidEnv) throw new Error(`Environment not supported: ${env}`)

    return env
  }

  private getApiCredentials = (): EnvironmentSecrets => {
    const clientId = process.env["API_CLIENT_ID"]
    const clientSecret = process.env["API_CLIENT_SECRET"]

    return {
      clientId,
      clientSecret
    }
  }

  getAuthParams(): Record<string, string> {
    return {
      client_id: this.clientId,
      redirect_uri: this.callbackUrl,
      state: this.state,
      response_type: "code"
    }
  }

  // https://internal-dev.api.service.nhs.uk/oauth2-mock/authorize
  // ?client_id=9AfEOqUltvbzj8YKXPZmN1ZfwaCRo4hs&redirect_uri=https://example.org/callback&state=123&response_type=code
  getAuthUrl(): string {
    return `/authorize`
  }

  getTokenUrl(): string {
    return `/token`
  }
}

export const getAuthForm = async (axiosInstance: AxiosInstance, requestUrl: string): Promise<AxiosResponse> => {
  const response = await axiosInstance.get(requestUrl)

  if (response.status !== 200) {
    throw new Error(`Cannot retrieve auth token: ${requestUrl} returned ${response.data}`)
  }

  return response
}

type FormInputs = {
  username?: string
  login?: string
}

type FormModel = {
  action: string
  method: string
  data: FormInputs
}

export const parseAuthForm = (htmlForm: string): FormModel => {
  const {document} = new JSDOM(htmlForm).window
  const form = document.getElementById("kc-form-login")

  const formAction = form.getAttribute("action")
  const formMethod = form.getAttribute("method")
  const formInputs = form.getElementsByTagName("input")

  const inputs: FormInputs = {}
  for (const item of formInputs) {
    const name: string = item.getAttribute("name")
    const value: string = item.getAttribute("value")
    inputs[name] = value
  }

  // Set username with test user value, e.g. 656005750108
  inputs["username"] = "656005750108"

  return {
    action: formAction,
    method: formMethod,
    data: inputs
  }
}

/**
 *
 * @param url - the url to the /token endpoint
 * @param formData - key value map of the form inputs
 * @returns Apigee auth token
 */
export const sendAuthForm = async (axiosInstance: AxiosInstance, form: FormModel): Promise<AxiosResponse> => {
  const response = await axiosInstance.post(
    form.action,
    form.data,
    // {username: "656005750108"},
    {
      headers: {
        "Content-Type": "application/x-www-form-urlencoded"
      }
    }
  )

  //NOSONAR
  // if (result.status !== 200) throw `Could not retrieve token: ${result.data}`

  return response
}

export const getAxiosInstance = (client: AuthClient): AxiosInstance => {
  const jar = new CookieJar()
  const axiosInstance = wrapper(
    axios.create({
      jar,
      // withCredentials: true,
      baseURL: client.getBaseUrl()
    })
  )

  return axiosInstance
}

type AuthCallbackResponseData = {
  code: string
  state: string
}

const parseAuthCallbackResponse = (authResponse: AxiosResponse): AuthCallbackResponseData => {
  const responsePath: string = authResponse.request.path // /?code=UmsJVKNA&state=1234567890

  const searchParams = new URLSearchParams(responsePath.substring(2)) // skip the leading `/?`
  return {
    code: searchParams.get("code"),
    state: searchParams.get("state")
  }
}

type TokenResponse = {
  access_token: string
  expires_in: string
  refresh_token: string
  refresh_token_expires_in: string
  refresh_count: string
  token_type: string
  sid: string
}

const exchangeCodeForToken = async (
  axiosClient: AxiosInstance,
  authClient: AuthClient,
  authCallbackData: AuthCallbackResponseData
): Promise<TokenResponse> => {
  const data = {
    grant_type: "authorization_code",
    code: authCallbackData.code,
    client_id: authClient.clientId,
    client_secret: authClient.clientSecret,
    redirect_uri: authClient.callbackUrl
  }

  const tokenResponse = await axiosClient.post(authClient.getTokenUrl(), data, {
    headers: {
      "Content-Type": "application/x-www-form-urlencoded"
    }
  })

  return tokenResponse.data
}

export const getAuthToken = async (): Promise<string> => {
  const client = new AuthClient()
  const axiosInstance = getAxiosInstance(client)

  const query = Object.entries(client.getAuthParams())
    .map(([key, value]) => `${key}=${value}`)
    .join("&")
  const authPath = `${client.getAuthUrl()}?${query}`

  const authFormResponse = await getAuthForm(axiosInstance, authPath)
  const authFormData = parseAuthForm(authFormResponse.data)

  const authFormSubmissionResponse = await sendAuthForm(axiosInstance, authFormData)
  const callbackData = parseAuthCallbackResponse(authFormSubmissionResponse)

  const tokenResponse = await exchangeCodeForToken(axiosInstance, client, callbackData)

  return tokenResponse.access_token
}
