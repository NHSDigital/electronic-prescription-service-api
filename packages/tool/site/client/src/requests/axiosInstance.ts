import axios, {InternalAxiosRequestConfig} from "axios"
import {v4 as uuidv4} from "uuid"

const axiosInstance = axios.create({
  validateStatus: () => true
})

axiosInstance.interceptors.request.use(
  (config: InternalAxiosRequestConfig) => {
    const requestId = uuidv4()
    config.headers["x-correlation-id"] = requestId

    return config
  },
  (error) => {
    return Promise.reject(Error(error))
  }
)

export {axiosInstance}
