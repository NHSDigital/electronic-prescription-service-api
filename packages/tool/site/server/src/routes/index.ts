import statusRoutes from "./health/get-status"
import loginRoute from "./auth/login"
import oauthCallbackRoute from "./auth/callback"
import refreshRoute from "./auth/refresh"
import logoutRoute from "./auth/logout"
import configRoutes from "./config/config"
import sessionRoutes from "./state/session"
import editRoutes from "./prescribe/edit"
import resetRoutes from "./prescribe/reset"
import signRoutes from "./sign/uploadSignatures"
import sendRoutes from "./sign/downloadSignatures"
import cancelRoutes from "./prescribe/cancel"
import validatorRoutes from "./validate/validator"
import searchRoutes from "./tracker/tracker"
import releaseRoutes from "./dispense/release"
import returnRoutes from "./dispense/return"
import dispenseRoutes from "./dispense/dispense"
import claimRoutes from "./dispense/claim"
import withdrawRoutes from "./dispense/withdraw"
import comparePrescriptions from "./api/comparePrescriptions"
import convertRoutes from "./api/convert"
import sendPrescriptions from "./api/send"
import doseToTextRoutes from "./dose-to-text"
import {isSandbox} from "../services/environment"
import {CONFIG} from "../config"
import {ServerRoute} from "@hapi/hapi"

const authRoutes: Array<ServerRoute> = [
  loginRoute,
  oauthCallbackRoute,
  refreshRoute,
  logoutRoute
]

const apiRoutes: Array<ServerRoute> = [
  ...comparePrescriptions,
  ...convertRoutes,
  ...sendPrescriptions
]

const stateRoutes: Array<ServerRoute> = [
  ...sessionRoutes
]

const prescribingRoutes: Array<ServerRoute> = [
  ...editRoutes,
  ...resetRoutes,
  ...signRoutes,
  ...sendRoutes,
  ...cancelRoutes
]

const validateRoutes: Array<ServerRoute> = [
  ...validatorRoutes
]

const dispensingRoutes: Array<ServerRoute> = [
  ...releaseRoutes,
  ...returnRoutes,
  ...dispenseRoutes,
  ...claimRoutes,
  ...withdrawRoutes
]

const trackerRoutes: Array<ServerRoute> = [
  ...searchRoutes
]

const healthcheckRoutes: Array<ServerRoute> = [
  ...statusRoutes
]

const routes: Array<ServerRoute> = isSandbox(CONFIG.environment)
  ? [
    ...healthcheckRoutes,
    ...doseToTextRoutes,
    oauthCallbackRoute
  ]
  : [
    ...configRoutes,
    ...authRoutes,
    ...apiRoutes,
    ...stateRoutes,
    ...healthcheckRoutes,
    ...prescribingRoutes,
    ...validateRoutes,
    ...dispensingRoutes,
    ...trackerRoutes,
    ...doseToTextRoutes
  ]

export default routes
