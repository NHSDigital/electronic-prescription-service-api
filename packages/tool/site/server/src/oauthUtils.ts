import ClientOAuth2 from "client-oauth2"
import {CONFIG} from "./config"
import {getRegisteredCallbackUrl} from "./routes/helpers"
import Hapi from "@hapi/hapi"
import {URLSearchParams} from "url"
import axios from "axios"
import * as jsonwebtoken from "jsonwebtoken"

//TODO: handle system oauth

function createOAuthClient(): ClientOAuth2 {
  return new ClientOAuth2({
    clientId: CONFIG.apigeeAppClientId,
    clientSecret: CONFIG.apigeeAppClientSecret,
    redirectUri: getRegisteredCallbackUrl("callback"),
    accessTokenUri: `${CONFIG.apigeeEgressHost}/oauth2/token`,
    authorizationUri: `${CONFIG.publicApigeeHost}/oauth2/authorize`,
    body: {
      client_id: CONFIG.apigeeAppClientId,
      client_secret: CONFIG.apigeeAppClientSecret
    }
  })
}

// TODO: refresh token
export async function refreshToken(data: ClientOAuth2.Data): Promise<ClientOAuth2.Token> {
  const oauthClientToken = createOAuthClient().createToken(data)
  const refreshedToken = await oauthClientToken.refresh()
  return refreshedToken
}

export interface OAuthTokenResponse {
  access_token: string
  refresh_token: string
  scope: string
  token_type: string
  expires_in: number
}

export interface UserInfoResponse {
  sub: string,
  nationalrbacaccess: {
    nhsid_nrbac_roles: Array<{
      person_roleid: string
    }>
  }
}

interface CIS2TokenResponse extends OAuthTokenResponse {
  id_token: string
}

export async function getCIS2TokenFromAuthCode(request: Hapi.Request): Promise<CIS2TokenResponse> {
  const authorisationCode = request.query.code

  const bodyParams = new URLSearchParams({
    grant_type: "authorization_code",
    client_id: CONFIG.cis2AppClientId,
    client_secret: CONFIG.cis2AppClientSecret,
    redirect_uri: "https://int.api.service.nhs.uk/eps-api-tool/callback",
    code: authorisationCode
  })
  const axiosCIS2TokenResponse = await axios.post<CIS2TokenResponse>(
    `https://${CONFIG.cis2EgressHost}/openam/oauth2/realms/root/realms/NHSIdentity/realms/Healthcare/access_token`,
    bodyParams
  )

  return axiosCIS2TokenResponse.data
}

export async function exchangeCIS2IdTokenForApigeeAccessToken(idToken: string): Promise<OAuthTokenResponse> {
  const apiKey = CONFIG.apigeeAppClientId
  const privateKey = CONFIG.apigeeAppJWTPrivateKey
  const audience = `${CONFIG.publicApigeeHost}/oauth2/token`
  const keyId = CONFIG.apigeeAppJWTKeyId

  const jwt = jsonwebtoken.sign({}, Buffer.from(privateKey, "base64").toString("utf-8"), {
    algorithm: "RS512",
    issuer: apiKey,
    subject: apiKey,
    audience: audience,
    keyid: keyId,
    expiresIn: 300,
    jwtid: crypto.randomUUID()
  })

  //Token Exchange for OAuth2 Access Token
  const bodyParams = new URLSearchParams({
    subject_token: idToken,
    client_assertion: jwt,
    subject_token_type: "urn:ietf:params:oauth:token-type:id_token",
    client_assertion_type: "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
    grant_type: "urn:ietf:params:oauth:grant-type:token-exchange"
  })

  // TODO: /token failure
  // eslint-disable-next-line max-len
  const axiosOAuthTokenResponse = await axios.post<OAuthTokenResponse>(`${CONFIG.apigeeEgressHost}/oauth2/token`, bodyParams)
  return axiosOAuthTokenResponse.data
}

// eslint-disable-next-line max-len
export async function getApigeeAccessTokenFromAuthCode(request: Hapi.Request, mock: boolean): Promise<OAuthTokenResponse> {
  // TODO: handle code not present
  const authorisationCode = request.query.code

  // TODO: match state on request

  const bodyParams = new URLSearchParams({
    grant_type: "authorization_code",
    client_id: CONFIG.apigeeAppClientId,
    client_secret: CONFIG.apigeeAppClientSecret,
    redirect_uri: getRegisteredCallbackUrl("callback"),
    code: authorisationCode
  })
  const path = mock ? "oauth2-mock/token" : "oauth2/token"
  // TODO: /token failure
  const axiosOAuthTokenResponse = await axios.post<OAuthTokenResponse>(`${CONFIG.apigeeEgressHost}/${path}`, bodyParams)
  return axiosOAuthTokenResponse.data
}

export function getSelectedRoleFromTokenResponse(tokenResponse: OAuthTokenResponse): string | undefined {
  const decodedToken = jsonwebtoken.decode(tokenResponse.access_token) as jsonwebtoken.JwtPayload

  return decodedToken?.["selected_roleid"] as string | undefined
}

export async function getUserInfoRbacRoleFromCIS2Token(
  request: Hapi.Request,
  cis2Token: CIS2TokenResponse
): Promise<string> {
  const logger = request.logger

  const access_token = cis2Token.access_token
  const id_token = cis2Token.id_token

  logger.info("Fetching userinfo to verify sub claim and get RBAC role")
  const axiosUserInfoResponse = await axios.get<UserInfoResponse>(
    `https://${CONFIG.cis2EgressHost}/openam/oauth2/realms/root/realms/NHSIdentity/realms/Healthcare/userinfo`,
    {
      headers: {
        Authorization: `Bearer ${access_token}`
      }
    }
  )
  logger.info(`Userinfo response: ${JSON.stringify(axiosUserInfoResponse.data)}`)

  // Verify sub claim
  const decodedIdToken = jsonwebtoken.decode(id_token) as jsonwebtoken.JwtPayload
  if (decodedIdToken.sub !== axiosUserInfoResponse.data.sub) {
    throw new Error(
      `sub claim in id_token does not match sub in userinfo response:`
        +`${decodedIdToken.sub} !== ${axiosUserInfoResponse.data.sub}`
    )
  }

  const roles = axiosUserInfoResponse.data.nationalrbacaccess.nhsid_nrbac_roles
  logger.info(`Userinfo roles: ${JSON.stringify(roles)}`)

  if (roles.length === 0) {
    throw new Error("No roles found for user")
  }

  logger.info(`Selecting first role from userinfo roles: ${JSON.stringify(roles[0])}`)
  return roles[0].person_roleid
}
