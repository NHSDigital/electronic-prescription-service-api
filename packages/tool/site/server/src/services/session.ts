import Hapi from "@hapi/hapi"
import {getUtcEpochSeconds} from "../routes/util"
import {CONFIG} from "../config"
import {isLocal} from "./environment"
import {OAuthTokenResponse} from "../oauthUtils"

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function getSessionValue(key: string, request: Hapi.Request): any {
  const sessionValue = request.yar.get(key)
  if (isLocal(CONFIG.environment)) {
    if (sessionValue === null) {
      request.logger.error(`Failed to retrieve session value for key: ${key}`)
    } else {
      request.logger.info(`Retrieved ${key} from session with value: ${JSON.stringify(sessionValue)}`)
    }
  }
  if (sessionValue && Object.keys(sessionValue).length === 1 && Object.keys(sessionValue)[0] === "arrayValues") {
    return sessionValue.arrayValues
  }
  return sessionValue
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function getSessionValueOrDefault(key: string, request: Hapi.Request, fallback: unknown): any {
  return getSessionValue(key, request) ?? fallback
}

export function setSessionValue(key: string, value: unknown, request: Hapi.Request): void {
  if (Array.isArray(value)) {
    value = {arrayValues: value}
  }
  if (isLocal(CONFIG.environment)) {
    request.logger.info(`Saving ${key} to session with value: ${JSON.stringify(value)}`)
  }
  request.yar.set(key, value)
}

export function appendToSessionValue(key: string, value: unknown, request: Hapi.Request): void {
  const existingValue = getSessionValueOrDefault(key, request, [])
  if (!Array.isArray(existingValue)) {
    throw Error(`Cannot append to session value with key: '${key}', session value is not an array`)
  }
  const mergedValue = existingValue.concat(value)
  setSessionValue(key, mergedValue, request)
}

export function appendToSessionValueWithoutDuplication(key: string, value: unknown, request: Hapi.Request): void {
  const existingValue = getSessionValueOrDefault(key, request, [])
  if (!Array.isArray(existingValue)) {
    throw Error(`Cannot append to session value with key: '${key}', session value is not an array`)
  }
  if (!existingValue.includes(value)) {
    const mergedValue = existingValue.concat(value)
    setSessionValue(key, mergedValue, request)
  }
}

export function removeFromSessionValue(key: string, value: unknown, request: Hapi.Request): void {
  const existingValue = getSessionValueOrDefault(key, request, [])
  if (!Array.isArray(existingValue)) {
    throw Error(`Cannot remove an entry from session value with key: '${key}', session value is not an array`)
  }
  const valueWithEntryRemoved = existingValue.filter(v => v !== value)
  setSessionValue(key, valueWithEntryRemoved, request)
}

export function clearSessionValue(key: string, request: Hapi.Request): void {
  request.yar.clear(key)
}

function createAuthSession(tokenResponse: OAuthTokenResponse, request: Hapi.Request, h: Hapi.ResponseToolkit) {
  request.cookieAuth.set({})
  const accessTokenFetchTime = getUtcEpochSeconds()
  const refreshTokenTimeout = CONFIG.refreshTokenTimeout
  const timeTillRefresh = 599
  const nextRefreshTime = accessTokenFetchTime + timeTillRefresh - 10
  request.cookieAuth.ttl(refreshTokenTimeout)
  // TODO: type session values
  setSessionValue("access_token_data", tokenResponse, request)
  setSessionValue("next_refresh_time", nextRefreshTime, request)
  h.state("Next-Refresh-Time", nextRefreshTime.toString(), {isHttpOnly: false})
  h.state("Access-Token-Fetched", accessTokenFetchTime.toString(), {isHttpOnly: false})
  h.state("Access-Token-Set", "true", {isHttpOnly: false, ttl: refreshTokenTimeout})
  h.state("Token-Expires-In", (refreshTokenTimeout / 1000).toString(), {isHttpOnly: false})
}

export function createSandboxAuthSession(request: Hapi.Request, h: Hapi.ResponseToolkit): void {
  const sandboxTokenResponse: OAuthTokenResponse = {
    access_token: "sandboxAccessToken",
    refresh_token: "sandboxRefreshToken",
    scope: "",
    token_type: "Bearer",
    expires_in: 1000 * 60 * 60
  }
  createAuthSession(sandboxTokenResponse, request, h)
}

export function createCombinedAuthSession(
  tokenResponse: OAuthTokenResponse, request: Hapi.Request, h: Hapi.ResponseToolkit
): void {
  createAuthSession(tokenResponse, request, h)
  h.state("Auth-Method", "Combined")
  h.state("Auth-Level", "User")
}

export function createSeparateAuthSession(
  tokenResponse: OAuthTokenResponse, request: Hapi.Request, h: Hapi.ResponseToolkit
): void {
  createAuthSession(tokenResponse, request, h)
  h.state("Auth-Method", "Separate")
  h.state("Auth-Level", "User")
}

export function getApigeeAccessTokenFromSession(request: Hapi.Request): string {
  if (isLocal(CONFIG.environment)) {
    return "sandbox-token"
  }
  const accessTokenData = getSessionValue("access_token_data", request)
  return accessTokenData.access_token
}

export function clearSession(request: Hapi.Request, h: Hapi.ResponseToolkit): void {
  request.yar.reset()
  request.cookieAuth.clear()
  h.unstate("Next-Refresh-Time")
  h.unstate("Access-Token-Fetched")
  h.unstate("Access-Token-Set")
  h.unstate("Token-Expires-In")
  h.unstate("Auth-Method")
  h.unstate("Auth-Level")
}
