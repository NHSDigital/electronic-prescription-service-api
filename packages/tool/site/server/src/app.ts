import Hapi, {RouteDefMethods, RouteOptions} from "@hapi/hapi"
import routes from "./routes"
import HapiPino from "hapi-pino"
import Vision from "@hapi/vision"
import * as inert from "@hapi/inert"
import Yar from "@hapi/yar"
import Cookie from "@hapi/cookie"
import {isLocal, isSandbox} from "./services/environment"
import axios from "axios"
import {CONFIG} from "./config"
import {getSessionValue} from "./services/session"
import * as XLSX from "xlsx"

const init = async () => {
  axios.defaults.validateStatus = () => true

  const server = createServer()

  await registerAuthentication(server)
  await registerSession(server)
  await registerLogging(server)
  await registerStaticRouteHandlers(server)
  await registerViewRouteHandlers(server)

  addStaticRoutes(server)
  addDownloadRoutes(server)
  addApiRoutes(server)
  addViewRoutes(server)

  await server.start()
  server.log("info", `Server running on ${server.info.uri}`)
}

process.on("unhandledRejection", err => {
  console.log(err)
  process.exit(1)
})

function createServer() {
  return Hapi.server({
    port: 9000,
    host: "0.0.0.0",
    routes: {
      cors: true
    }
  })
}

async function registerAuthentication(server: Hapi.Server) {
  await server.register(Cookie)
  server.auth.strategy("session", "cookie", {
    cookie: {
      name: "auth",
      password: CONFIG.sessionKey,
      isSecure: true
    },
    redirectTo: () => {
      const needsLogin = !(isSandbox(CONFIG.environment) || isLocal(CONFIG.environment))
      return needsLogin ? `${CONFIG.baseUrl}login` : `${CONFIG.baseUrl}callback`
    }
  })
  server.auth.default("session")
}

async function registerSession(server: Hapi.Server) {
  await server.register({
    plugin: Yar,
    options: {
      storeBlank: true,
      // Use "0" maxCookieSize to force all session data to be written to cache
      maxCookieSize: 0,
      cache: {
        expiresIn: 24 * 60 * 60 * 1000
      },
      cookieOptions: {
        password: CONFIG.sessionKey,
        isSecure: true,
        isSameSite: "None"
      }
    }
  })
}

async function registerLogging(server: Hapi.Server) {
  await HapiPino.register(server, {
    // For non-local environments, dont pretty print to avoid spamming logs
    ...(isLocal(CONFIG.environment) && {
      transport: {
        target: "pino-pretty",
        options: {
          colorize: true,
          minimumLevel: "info",
          levelFirst: true,
          messageFormat: true,
          timestampKey: "time",
          translateTime: true,
          singleLine: false,
          mkdir: true,
          append: true
        }
      }
    }),
    // Redact Authorization headers, see https://getpino.io/#/docs/redaction
    redact: ["req.headers.authorization"],
    wrapSerializers: false
  })
}

async function registerStaticRouteHandlers(server: Hapi.Server) {
  await server.register(inert)
}

async function registerViewRouteHandlers(server: Hapi.Server) {
  await server.register(Vision)
  server.views({
    engines: {
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      html: require("handlebars")
    },
    relativeTo: __dirname,
    path: "templates"
  })
}

function addStaticRoutes(server: Hapi.Server) {
  server.route({
    method: "GET",
    path: `/static/{param*}`,
    options: {
      auth: false
    },
    handler: {
      directory: {
        path: "static"
      }
    }
  })
}

function addDownloadRoutes(server: Hapi.Server) {
  server.route({
    method: "GET",
    path: "/download/exception-report",
    handler: downloadExceptionReport
  })

  function downloadExceptionReport(request: Hapi.Request, h: Hapi.ResponseToolkit) {
    const exceptions = getSessionValue("exception_report", request)
    const fileName = "exception-report"
    const wb = XLSX.utils.book_new()
    const ws = XLSX.utils.json_to_sheet(exceptions)
    XLSX.utils.book_append_sheet(wb, ws, "Test Exception Report")

    return h
      .response(XLSX.write(wb, {type: "binary"}))
      .type("application/vnd.ms-excel")
      .header("content-disposition", `attachment; filename=${fileName}.xlsx;`)
      .encoding("binary")
      .code(200)
  }
}

function addApiRoutes(server: Hapi.Server) {
  server.route(routes)
}

function addViewRoutes(server: Hapi.Server) {
  server.route(addHomeView())

  if (isSandbox(CONFIG.environment)) {
    server.route(addView("dose-to-text"))
  } else {
    server.route(addView("login", true))
    server.route(addView("my-prescriptions"))
    server.route(addView("validate"))
    server.route(addView("compare-prescriptions"))
    server.route(addView("search"))
    server.route(addView("view"))
    server.route(addView("tracker"))
    server.route(addView("prescribe/load"))
    server.route(addView("prescribe/edit"))
    server.route(addView("prescribe/send", true))
    server.route(addView("prescribe/cancel"))
    server.route(addView("dispense/release"))
    server.route(addView("dispense/verify"))
    server.route(addView("dispense/return"))
    server.route(addView("dispense/dispense"))
    server.route(addView("dispense/withdraw"))
    server.route(addView("dispense/claim"))
  }

  if (isLocal(CONFIG.environment)) {
    server.route(addView("dose-to-text"))
  }

  server.route(addView("config"))

  function addHomeView(): Hapi.ServerRoute {
    return addView("/")
  }

  function addView(path: string, skipAuth?: boolean): Hapi.ServerRoute {
    const viewRoute = {
      method: "GET" as RouteDefMethods,
      path: path.startsWith("/") ? path : `/${path}`,
      handler: {
        view: {
          template: "index",
          context: {
            baseUrl: CONFIG.baseUrl,
            environment: CONFIG.environment
          }
        }
      }
    }

    if (skipAuth) {
      return {
        ...viewRoute,
        options: {
          auth: false
        } satisfies RouteOptions
      }
    }

    return viewRoute
  }
}

init()
