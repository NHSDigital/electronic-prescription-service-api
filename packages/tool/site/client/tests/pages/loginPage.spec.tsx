import {waitFor} from "@testing-library/react"
import {screen} from "@testing-library/dom"
import pretty from "pretty"
import * as React from "react"
import {AppContextValue} from "../../src"
import {renderWithContext} from "../renderWithContext"
import {Environment, int, internalDev} from "../../src/services/environment"
import LoginPage from "../../src/pages/loginPage"
import userEvent from "@testing-library/user-event"
import {redirect} from "../../src/browser/navigation"
import {axiosInstance} from "../../src/requests/axiosInstance"
import MockAdapter from "axios-mock-adapter"

const baseUrl = "baseUrl/"

const loginUrl = `${baseUrl}login`
const configUrl = `${baseUrl}getconfig`

const attendedAuthRedirectUrl = `https://attended-auth.com`
const unattendedAuthRedirectUrl = `https://unattended-auth.com`

jest.mock("../../src/browser/navigation")

const mock = new MockAdapter(axiosInstance)

beforeEach(() => {
  mock.reset()
  mock.onAny(configUrl).reply(200)
})
afterEach(() => mock.reset())

test("Displays user/system options in internal-dev", async () => {
  const container = await renderPage(internalDev)
  await waitFor(() => screen.getByText("Login"))
  expect(screen.getByText("Select access level:")).toBeTruthy()
  expect(screen.getByText("User - Mock")).toBeTruthy()
  expect(screen.getByText("System")).toBeTruthy()
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test("Redirects to attended simulated auth when selecting user access level in internal-dev", async () => {
  mock.onAny(loginUrl).reply(200, {
    redirectUri: attendedAuthRedirectUrl
  })

  const container = await renderPage(internalDev)
  await waitFor(() => screen.getByText("Login"))
  userEvent.click(screen.getByText("User - Mock"))
  await waitFor(() => screen.getByText("Redirecting to simulated auth..."))
  expect(pretty(container.innerHTML)).toMatchSnapshot()
  expect(redirect).toHaveBeenCalledWith(attendedAuthRedirectUrl)
})

test("Redirects to unattended auth when selecting system access level in internal-dev", async () => {
  mock.onAny(loginUrl).reply(200, {
    redirectUri: unattendedAuthRedirectUrl
  })

  const container = await renderPage(internalDev)
  await waitFor(() => screen.getByText("Login"))
  userEvent.click(screen.getByText("System"))
  expect(pretty(container.innerHTML)).toMatchSnapshot()
  await waitFor(() => expect(redirect).toHaveBeenCalledWith(unattendedAuthRedirectUrl))
})

test("Redirects to attended auth in integration", async () => {
  mock.onAny(loginUrl).reply(200, {
    redirectUri: attendedAuthRedirectUrl
  })

  const container = await renderPage(int)
  await waitFor(() => screen.getByText("Login"))
  await waitFor(() => screen.getByText("Redirecting to auth..."))
  expect(pretty(container.innerHTML)).toMatchSnapshot()
  expect(redirect).toHaveBeenCalledWith(attendedAuthRedirectUrl)
})

async function renderPage(environment: Environment) {
  const context: AppContextValue = {baseUrl, environment}
  const {container} = renderWithContext(<LoginPage />, context)
  return container
}
