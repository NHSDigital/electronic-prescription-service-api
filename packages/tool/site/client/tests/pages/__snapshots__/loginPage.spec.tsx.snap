// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Displays user/system options in internal-dev 1`] = `
"<h1 class="nhsuk-label-wrapper"><label class="nhsuk-label nhsuk-label--xl">Login</label></h1><label class="nhsuk-label">When prompted enter the following User Id: 555086689106</label><label class="nhsuk-label">Select access level:</label>
<div class="sc-dVBluf duNEre"><button class="nhsuk-button" aria-disabled="false" type="submit">User - Mock</button><button class="nhsuk-button" aria-disabled="false" type="submit">System</button></div>"
`;

exports[`Redirects to attended auth in integration 1`] = `"<h1 class="nhsuk-label-wrapper"><label class="nhsuk-label nhsuk-label--xl">Login</label></h1><label class="nhsuk-label">Redirecting to auth...</label>"`;

exports[`Redirects to attended simulated auth when selecting user access level in internal-dev 1`] = `"<h1 class="nhsuk-label-wrapper"><label class="nhsuk-label nhsuk-label--xl">Login</label></h1><label class="nhsuk-label">Redirecting to simulated auth...</label>"`;

exports[`Redirects to unattended auth when selecting system access level in internal-dev 1`] = `
"<h1 class="nhsuk-label-wrapper"><label class="nhsuk-label nhsuk-label--xl">Login</label></h1><label class="nhsuk-label">When prompted enter the following User Id: 555086689106</label><label class="nhsuk-label">Select access level:</label>
<div class="sc-dVBluf duNEre"><button class="nhsuk-button" aria-disabled="false" type="submit">User - Mock</button><button class="nhsuk-button" aria-disabled="false" type="submit">System</button></div>"
`;
